### DISCUSSION TITLE
Handling large amounts of data

### DISCUSSION THREAD

**Discussion by kragoth** [salesforce/akita#712] [Q&A]
Hi,

So, after 18 months now of an app in production that uses akita heavily I want to discuss ways of dealing with large amounts of data.

Now my idea of what causes my problem is based on a hunch as I cannot prove that my theory is totally correct but... the evidence seems to suggest it might be. So, feel free to shoot this down. :)

We have built an app that must work offline. To do this we sync all data to the device that the user may need to perform their job for that day. This can be 50-150MB (I think I've seen this up to 300MB of data but... on average probably 100MB) of data (we have spatial data to deal with). 
But, what this means is that any change to any store within akita results in the full state needing to be re-written to indexdb.
Now, I *think* when this write occurs, the thread is blocked. So, the UI freezes until the persist operation is done. Now this is normally pretty quick but, when you have a lot of data it can be 2-5 seconds.

I have just completed a massive refactor of our app in which I have moved all akita operations to a web worker. This has made a huge difference for us as we no longer get UI freezing when updates occur.

While this does relieve the problems from a UI perspecrtive, the web worker also becomes non responsive while the persist operation is occurring. Which means the UI won't receive updates from queries etc until the web worker is finished persisting the state.

So.... are there any options/ideas on how to deal with this or improve the behaviour.

I toyed with the idea what about a database per store.... but, I guess the downside to this is you could lose data consistency if one write succeeds and another fails.

Anyway, I'd be interested in hearing from anyone else that might be dealing with a lot of data.
Or maybe I'm totally off on what I think is causing the problem!

**Comment by NetanelBasal** [comment]
Hi!

First, I want to say that your issue isn't related to Akita. It will happen with any other solution. 

I'd try these options:
1. Save only the deltas in indexdb ( I'm not familiar with your application so I couldn't suggest more than that)
2. Save the state in these events: `offline`, `visiblityChange`, or use `requestIdleCallback` to save the state.

**Reply by kragoth** [reply]
Hey @NetanelBasal, it is my understanding that I have no control over how the data is being saved into indexdb. That is entirely managed by akita. From what I can tell by looking at the akita codebase, the entire state is saved to indexdb every time a change is made, not just the deltas. (Or do I have this entirely wrong?)

The second option is interesting.... and I think in theory would work (but, I don't think I can ask akita to only perform the flush to indexdb based on those events). However, this could lead to a situation where if the browser is forcefully closed or crashes that data would be lost. So, that pretty much rules that out.

It is entirely possible that there is no real solution to this problem, and to be fair I have got our app working were I'm not 95% happy with how it performs. I was just curious as to whether or not there was more I could be doing.

**Reply by NetanelBasal** [reply]
You don't have to use the built-in storage solution we provide. You can easily create your own implementation. 

You can use the snapshot manager and save everything to the db.

**Reply by NetanelBasal** [reply]
> However, this could lead to a situation where if the browser is forcefully closed or crashes, that data would be lost.

Try to check if there is a solution to this situation. My second suggestion is the most efficient way.

**Reply by kragoth** [reply]
"You can use the snapshot manager and save everything to the db." 
Sorry, could you explain that a bit for me? I'm not sure I understand what you mean by that.

With regards to implementing my own storage solution, that's an interesting idea. I would need to go learn a lot about that before I'd think about tackling it that way!

**Reply by NetanelBasal** [reply]
https://datorama.github.io/akita/docs/enhancers/snapshot

**Reply by kragoth** [reply]
Sorry, I guess I'm not following how the snapshot helps me in this scenario. I might be missunderstanding something.

**Reply by NetanelBasal** [reply]
That's related to my second suggestion. I meant that you can use indexdb ( without Akita ) and save the entire stores snapshot by using this functionality. But only when the suggested events occurred.

**Reply by kragoth** [reply]
Oooh right! I get it now.
So, I would implment my own "storage solution" that just uses the snapshot manager to push into indexdb when one of those events fires etc.

Hmm, interesting idea. this actually might work because there are some stores that have a really large amount of read only data. So, I could split them off into their own indexdb... that alone would probably be enough. 

But, I'm right in assuming that to split my read only data out of the same indexdb I would have to implement my own storage solution yeah? There's no built in mechanism for that?

**Reply by NetanelBasal** [reply]
I'm not sure I understand the question. See if this what you need - https://datorama.github.io/akita/docs/enhancers/persist-state#separate-branch-in-storage

**Reply by kragoth** [reply]
That might be what I'm after. This is where my lack of understanding of how some things work is showing 😛 .

So maybe I'll try use an example.

So lets say I have entities A, B, C

In my current set up if I make a change an entity of type A, akita will update it's in memory state which contains all entities of A, B and C, then it will flush that state to indexdb. (At least I think this is how it works).


If I swap to using the separate branch in storage strategy and I have this setup.
Entities A and B use branch YY
Entity C uses branch ZZ

When I make to entity of type A what happens?
Is it any different from my scenario above?
Or does only the branch YY data get flushed to index db?

Or am I totally wrong on how this works? 😛

**Reply by NetanelBasal** [reply]
Entity === Store?

**Reply by kragoth** [reply]
Oh yes, sorry, they are all entity stores.

**Reply by NetanelBasal** [reply]
When you update any store, Akita saves the entire application state to storage ( the stores that you provided ). If you split it, in your example, if you update entity C, it'll save only the entity C state to storage.

**Reply by kragoth** [reply]
oooh ok!
Well maybe that's all I need to do!

Thanks @NetanelBasal I'll see if I can get that working. Really appreciate the discussion. 😄

**Reply by NetanelBasal** [reply]
Sure, good luck.

**Reply by kragoth** [reply]
@NetanelBasal, just thought I'd let you know.
I have implemented the separate branch in storage strategy and it has made a noticible improvement!
Thanks again!

**Reply by NetanelBasal** [reply]
Great!
